bullet/src/ShapeFeatures.cc

```diff
@@ -0,0 +1,120 @@
+ *
+*/
+
+#include "ShapeFeatures.hh"
+#include <BulletCollision/Gimpact/btGImpactShape.h>
+
+#include <memory>
+
+namespace ignition {
+namespace physics {
+namespace bullet {
+
+/////////////////////////////////////////////////
+Identity ShapeFeatures::AttachMeshShape(
+    const Identity &_linkID,
+    const std::string &_name,
+    const ignition::common::Mesh &_mesh,
+    const Pose3d &_pose,
+    const LinearVector3d &_scale)
+{
+  double *vertices = nullptr;
+  int *indices = nullptr;
+  _mesh.FillArrays(&vertices, &indices);
+
+  unsigned int numVertices = _mesh.VertexCount();
+  unsigned int numIndices = _mesh.IndexCount();
+
+  const auto mTriMesh = std::make_shared<btTriangleMesh>();
+
+  for (unsigned int j = 0;  j < numVertices; ++j)
+  {
+    // Scale the vertex data
+    vertices[j*3+0] = vertices[j*3+0] * _scale[0];
+    vertices[j*3+1] = vertices[j*3+1] * _scale[1];
+    vertices[j*3+2] = vertices[j*3+2] * _scale[2];
+  }
+
+  // Create the Bullet trimesh
+  for (unsigned int j = 0; j < numIndices; j += 3)
+  {
+    btVector3 bv0(vertices[indices[j]*3+0],
+                  vertices[indices[j]*3+1],
+                  vertices[indices[j]*3+2]);
+
+    btVector3 bv1(vertices[indices[j+1]*3+0],
+                  vertices[indices[j+1]*3+1],
+                  vertices[indices[j+1]*3+2]);
+
+    btVector3 bv2(vertices[indices[j+2]*3+0],
+                  vertices[indices[j+2]*3+1],
+                  vertices[indices[j+2]*3+2]);
+
+    mTriMesh.get()->addTriangle(bv0, bv1, bv2);
+  }
+
+  auto gimpactMeshShape =
+    std::make_shared<btGImpactMeshShape>(mTriMesh.get());
+  gimpactMeshShape.get()->updateBound();
+
+  // TODO(lobotuerk) Save collision if needed
+  // collision->shape = gimpactMeshShape;
+
+  delete [] vertices;
+  delete [] indices;
+
+  const auto &linkInfo = this->links.at(_linkID);
+  const auto &modelID = linkInfo->model;
+  const auto &body = linkInfo->link.get();
+
+  auto poseWithInertia =
+    linkInfo->inertialPose.Inverse() * ignition::math::eigen3::convert(_pose);
+  const auto poseIsometry = ignition::math::eigen3::convert(poseWithInertia);
+  const auto poseTranslation = poseIsometry.translation();
+  const auto poseLinear = poseIsometry.linear();
+  btTransform baseTransform;
+  baseTransform.setOrigin(convertVec(poseTranslation));
+  baseTransform.setBasis(convertMat(poseLinear));
+
+  /* TO-DO(Lobotuerk): figure out if this line is needed */
+  // gimpactMeshShape->setMargin(btScalar(0.001));
+
+  dynamic_cast<btCompoundShape *>(
+    body->getCollisionShape())->addChildShape(
+    baseTransform, gimpactMeshShape.get());
+
+  auto identity = this->AddCollision(
+    _linkID, {_name, gimpactMeshShape, _linkID, modelID,
+    ignition::math::eigen3::convert(_pose), true});
+  return identity;
+
+}
+
+/////////////////////////////////////////////////
+Identity ShapeFeatures::CastToMeshShape(
+    const Identity &_shapeID) const
+{
+  if (this->collisions.at(_shapeID)->isMesh)
```

Comment:
Should we check that there's a valid pointer there before calling `isMesh`?

Reply:
`unordered_map.at()` does not return a non valid pointer AFAIK, its a valid or it throws and exception

Reply:
https://github.com/ignitionrobotics/ign-physics/pull/208#discussion_r623903046 This applies here too